presto-main/src/main/java/com/facebook/presto/operator/spiller/BinaryFileSpiller.java

```diff
@@ -0,0 +1,152 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.spiller;
+
+import com.facebook.presto.block.PagesSerde;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.BlockEncodingSerde;
+import com.facebook.presto.spi.spiller.Spiller;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.common.io.Closer;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import io.airlift.concurrent.MoreFutures;
+import io.airlift.slice.InputStreamSliceInput;
+import io.airlift.slice.OutputStreamSliceOutput;
+import io.airlift.slice.RuntimeIOException;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Stream;
+
+import static com.facebook.presto.spi.StandardErrorCode.INTERNAL_ERROR;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class BinaryFileSpiller
+        implements Spiller
+{
+    private static final Path SPILL_PATH = Paths.get("/tmp/spills");
+
+    static {
+        SPILL_PATH.toFile().mkdirs();
+    }
+
+    private final Path targetDirectory;
+    private final Closer closer = Closer.create();
+    private final BlockEncodingSerde blockEncodingSerde;
+
+    private int spillsCount = 0;
+    private final ListeningExecutorService executor;
+
+    public BinaryFileSpiller(BlockEncodingSerde blockEncodingSerde, ListeningExecutorService executor)
+    {
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.executor = requireNonNull(executor, "executor is null");
+        try {
+            this.targetDirectory = Files.createTempDirectory(SPILL_PATH, "presto-spill");
+        }
+        catch (IOException e) {
+            throw new PrestoException(INTERNAL_ERROR, "Failed to create spill directory", e);
+        }
+    }
+
+    @Override
+    public CompletableFuture<?> spill(Iterator<Page> pageIterator)
+    {
+        Path spillPath = getPath(spillsCount++);
```

Comment:
Each `BinaryFileSpiller` instance is guaranteed to have it's own directory, so only concurrent calls to this `spill` method could result in race conditions. But this can not happen, since each `BinaryFileSpiller` is used only by one single operator thus by one thread.